### PR TITLE
Render Fn* trait objects and impl types as rust does

### LINKED ISSUE #3012: Type hints for `Fn*` traits don't use the usual syntax sugar
Type hints for `impl Fn<(Ts...)> Output = O>` don't use the usual syntax sugar of `impl Fn(Ts...) -> O`. The same is probably true for `impl FnMut`, impl `FnOnce`, `dyn Fn*` and potentially other similar things.

### PR DESCRIPTION
**Pull request by Veykril**
Also fixes raw ptr impl types being rendered ambiguously with multiple predicates.
This moves out the `FamousDefs::Fixture` into its own file as well, cause I figured it is big enough to get its own file at this point + we also get highlighting this way when editing it.

Fixes #3012

### CODE REVIEW

File: crates/hir_ty/src/display.rs [hunk 1]
```diff
@@ -257,25 +258,45 @@ impl HirDisplay for ApplicationTy {
...
+                    write!(f, "*{}", m.as_keyword_for_ptr())?;
                 } else {
-                    t.hir_fmt(f)?;
+                    write!(f, "&{}", m.as_keyword_for_ref())?;
+                }
+
+                let datas;
+                let predicates = match t {
+                    Ty::Dyn(predicates) if predicates.len() > 1 => {
+                        std::borrow::Cow::Borrowed(predicates.as_ref())
```
Comment (jonas-schievink) [inline crates/hir_ty/src/display.rs:275]: Importing `Cow` seems reasonable here

**Review by jonas-schievink — COMMENTED** [review]
> Also fixes raw ptr impl types being rendered ambiguously with multiple predicates.

Can you add a test for this too? Hard to imagine what this looks like

**Comment by Veykril** [comment]
It was already kind of tested by the inlay hints module, didn't realize there was a display test module for hir_ty, added that now https://github.com/rust-analyzer/rust-analyzer/pull/7260/files#diff-a917baaa08f60d032c715705b55ff8f2047af956c4b6446fc383c21043d7d5cd

It was emitting `*const impl Unpin + Sized` before which is wrong as it is ambiguous.

File: crates/ide_db/src/helpers/famous_defs_fixture.rs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    pub use crate::{
+        convert::From,
+        default::Default,
+        iter::{IntoIterator, Iterator},
+        ops::{Fn, FnMut, FnOnce},
+        option::Option::{self, *},
+    };
+}
+#[prelude_import]
+pub use prelude::*;
```
Comment (matklad) [inline crates/ide_db/src/helpers/famous_defs_fixture.rs:120]: :+1:

**Review by matklad — COMMENTED** [review]
bors r+

**Pull request merged**